README.md
Stock Additions 0.7.7
=====

Stock Additions is a fork of openpilot designed to be minimal in UI design while boasting various feature additions and behavior improvements over stock. I have a Toyota Corolla with comma pedal, so most of my changes are designed to improve the longitudinal performance.

Want to request a feature or create a bug report? [Open an issue here!](https://github.com/ShaneSmiskol/openpilot/issues/new/choose)

If you'd like to support my development of Stock Additions with a [dollar for a RaceTrac ICEE](https://paypal.me/ssmiskol) (my PayPal link). Thanks! 🥰

-----

Highlight Features
=====

* [**Dynamic follow (now with profiles!)**](#dynamic-follow-3-profiles) - 3 + auto profiles to control distance
  * [**`auto-df` model for automatic distance profile switching**](#Automatic-DF-profile-switching)
* **Lane Speed**
  * [**Lane Speed Alerts**](#Lane-Speed-alerts) - alerts for when an adjacent lane is faster
  * [**Dynamic camera offsetting**](#Dynamic-camera-offset-based-on-oncoming-traffic) - automatically moves you over if adjacent lane has oncoming traffic
* [**NEW❗ Curvature Learner v5**](#curvature-learner) - zorrobyte's curvature learner with some additions
* [**Dynamic gas**](#dynamic-gas) - smoother gas control
* [**PI > PID for longcontrol**](#Long-control-uses-a-PID-loop) - fix for overshoot with comma pedal
* [**Customize this fork**](#Customize-this-fork-opEdit) - easily edit fork parameters with support for live tuning
* [**Automatic updates**](#Automatic-updates)
* [**Offline crash logging**](#Offline-crash-logging) - check out `/data/community/crashes`

Documentation
=====
* [**Quick Installation**](#Quick-installation)
* [**Branches**](#Branches)
* [**Videos**](#Videos)

-----

Dynamic follow (3 profiles)
-----
Dynamic follow aims to provide the stock (Toyota) experience of having three different distance settings. Dynamic follow works by dynamically changing the distance in seconds which is sent to the long MPC to predict a speed to travel at. Basically, if the lead is decelerating or might soon, increase distance to prepare. And if the lead is accelerating, reduce distance to get up to speed quicker.

Dynamic follow works if openpilot can control your vehicle's gas and brakes (longitudinal). [Check if openpilot can control your vehicle's longitudinal from this list.](https://github.com/commaai/openpilot#supported-cars)

Just use the button on the button right of the screen while driving to change between these profiles:
  * [`traffic`](#Videos) - Meant to keep you a bit closer in traffic, hopefully reducing cut-ins. Always be alert, as you are with any driving assistance software.
  * `relaxed` - This is the default dynamic follow profile for casual driving.
  * `roadtrip` - This profile is for road trips where you're mainly on two lane highways and don't want to be following particularly closely; at night for example.
  * [`auto`](#Automatic-DF-profile-switching) - The auto dynamic follow model was trained on about an hour of me manually cycling through the different profiles based on driving conditions, this profile tries to replicate those decisions entirely on its own.

<img src=".media/df_profiles.jpg?raw=true">

Automatic DF profile switching
-----
I've trained a custom model with Keras that takes in the past 35 seconds of your speed, the lead's speed and the lead's distance. With these inputs, it tries to correctly predict which profile is the best for your current situation.

It's only been trained on about an hour of data, so it's not perfect yet, but it's great for users who just want to set it and forget it. **To enable the `auto` profile, simply tap the profile changing button for dynamic follow until it reaches the `auto` profile!**

If you're annoyed by the silent alerts that show when the model has changed the profile automatically, just use [opEdit](#Customize-this-fork-opEdit) and set `hide_auto_df_alerts` to `True`. Auto profile and model will remain functional but will not show alerts.

Resources:
- [The auto-df repo.](https://github.com/ShaneSmiskol/auto-df)
- [The model file.](https://github.com/ShaneSmiskol/openpilot/blob/stock_additions/selfdrive/controls/lib/dynamic_follow/auto_df.py)
- I converted the Keras model to be able to run with pure NumPy using [Konverter](https://github.com/ShaneSmiskol/Konverter).

Lane Speed alerts
-----
This feature alerts you of faster-travelling adjacent lanes and can be configured using the on-screen *LS* button on the bottom right to either be disabled, audible, or silent.

The idea behind this feature is since we often become very relaxed behind the wheel when being driven by openpilot, we don't always notice when we've become stuck behind a slower-moving vehicle. When either the left or right adjacent lane is moving faster than your current lane, LaneSpeed alerts the user that a faster lane is available so that they can make a lane change, overtaking the slower current lane. Thus saving time in the long run on long roadtrips or in general highway driving!

The original idea is thanks to [Greengree#5537](https://github.com/greengree) on Discord. This feature is available at 35 mph and up.

Dynamic camera offset (based on oncoming traffic)
-----
This feature automatically adjusts your position in the lane if an adjacent lane has oncoming traffic. For example, if you're on a two-lane highway and the left adjacent lane has oncoming cars, LaneSpeed recognizes those cars and applies an offset to your `CAMERA_OFFSET` to move you over in the lane, keeping you farther from oncoming cars.

**This feature is available from 35 to ~60 mph due to a limitation with the Toyota radar**. It may not recognize oncoming traffic above 60 mph or so. To enable or disable this feature, use `opEdit` and change this parameter: `dynamic_camera_offset`.

Curvature Learner
-----
From [zorrobyte](https://github.com/zorrobyte/openpilot), I've added his Curvature Learner to help fix hugging in curves with a few modifications of my own to improve the idea further.

***What's new in Curvature Learner v5:***
  - Moved to using the curvature from dPoly rather than steering angle for curve measurement and learning. This helps fix oscilation issues with slight curves.
    - Curvature is (rudimentarily) calculated as the lateral position of the dPoly path at 0.9 seconds into the future.
  - Added vehicle speed as a learning factor
  - Added curve direction as a learning factor (left and right curves learn different offsets)
  - And the biggest difference is the introduction of clustering, I've used sklearn's KMeans algorithm to find the most common clusters of data (speed vs road curvature on a 2D graph). The curvature learner will group the sample into the closest cluster and learn/use the offset from that cluster.

Special thanks to [Trae#2379](https://github.com/d412k5t412) for helping me gather data used to calculate the clusters!
<p align="center">
  <img src=".media/curvature-clusters.png" height="350">
</p>

Dynamic gas
-----
Dynamic gas aims to provide a smoother driving experience in stop and go traffic (under 20 mph) by reducing the maximum gas that can be applied based on your current velocity, the relative velocity of the lead, the acceleration of the lead, and the distance of the lead. This usually results in quicker and smoother acceleration from a standstill without the jerking you get in stock openpilot with comma pedal (ex. taking off from a traffic light). It tries to coast if the lead is just inching up, it doesn’t use maximum gas as soon as the lead inches forward. When you are above 20 mph, relative velocity and the current following distance in seconds is taken into consideration.

All cars that have a comma pedal are supported! However to get the smoothest acceleration, I've custom tuned gas curve profiles for the following cars:

pedal cars:
  * 2017 Toyota Corolla (non-TSS2)
  * Toyota RAV4 (non-TSS2)
  * 2017 Honda Civic
  * 2019 Honda Pilot

non-pedal cars:
  * None yet

If you have a car without a pedal, or you do have one but I haven't created a profile for you yet, please let me know and we can develop one for your car to test.

Long control uses a PID loop
-----
I've added a custom implementation of derivative to the PI loop controlling the gas and brake output sent to your car. Derivative (change in error) is calculated based on the current and last error and added to the class's integral variable. It's essentially winding down integral according to derivative. It helps fix overshoot on some cars with the comma pedal and increases responsiveness (like when going up and down hills) on all other cars! Still need to figure out the tuning, right now it's using the same derivative gain for all cars. Test it out and let me know what you think!

Derivative is disabled by default due to only one tune for all cars, but can be enabled by using [opEdit](#Customize-this-fork-opEdit) and setting the `enable_long_derivative` parameter to `True`. It works well on my '17 Corolla with pedal.

Customize this fork (opEdit)
-----
This is a handy tool to change your `opParams` parameters without diving into any json files or code. You can specify parameters to be used in any fork's operation that supports `opParams`. First, ssh in to your EON and make sure you're in `/data/openpilot`, then start `opEdit`:
```python
cd /data/openpilot
python op_edit.py  # or ./op_edit.py
```

[To see what features opEdit has, click me!](/OPEDIT_FEATURES.md)

Here are the main parameters you can change with this fork:
- **Tuning params**:
  - `camera_offset` **`(live!)`**: Your camera offset to use in lane_planner.py. Helps fix lane hugging
  - `steer_ratio` **`(live!)`**: The steering ratio you want to use with openpilot. If you enter None, it will use the learned steer ratio from openpilot instead
  - `enable_long_derivative`: This enables derivative-based integral wind-down to help overshooting within the PID loop. Useful for Toyotas with pedals or cars with bad long tuning
- **General fork params**:
  - `alca_nudge_required`: Whether to wait for applied torque to the wheel (nudge) before making lane changes
  - `alca_min_speed`: The minimum speed allowed for an automatic lane change
  - `upload_on_hotspot`: Controls whether your EON will upload driving data on your phone's hotspot
  - `no_ota_updates`: Set this to True to disable all automatic updates. Reboot to take effect
  - `disengage_on_gas`: Whether you want openpilot to disengage on gas input or not
- **Dynamic params**:
  - `dynamic_gas`: Whether to use [dynamic gas](#dynamic-gas) if your car is supported
  - `global_df_mod` **`(live!)`**: The multiplier for the current distance used by dynamic follow. The range is limited from 0.85 to 1.2. Smaller values will get you closer, larger will get you farther. This is applied to ALL profiles!
  - `min_TR` **`(live!)`**: The minimum allowed following distance in seconds. Default is 0.9 seconds, the range of this mod is limited from 0.85 to 1.3 seconds. This is applied to ALL profiles!
  - `hide_auto_df_alerts`: Hides the alert that shows what profile the model has chosen
  - [`dynamic_camera_offset`](#Dynamic-camera-offset-based-on-oncoming-traffic): Whether to automatically keep away from oncoming traffic. Works from 35 to ~60 mph
  - [`dynamic_camera_offset_time`](#Dynamic-camera-offset-based-on-oncoming-traffic): How long to keep the offset after losing the oncoming lane/radar track in seconds
  - `dynamic_follow`: *Deprecated, use the on-screen button to change profiles*
- **Experimental params**:
  - `support_white_panda`: This allows users with the original white panda to use openpilot above 0.7.7. The high precision localizer's performance may be reduced due to a lack of GPS
  - `prius_use_lqr`: If you have a newer Prius with a good angle sensor, you can try enabling this to use LQR
  - `corolla_use_lqr`: Enable this to use LQR for lat with your Corolla (2017) *(can be enabled for all years by request)*

A full list of parameters that you can modify are [located here](common/op_params.py#L40).

An archive of opParams [lives here.](https://github.com/ShaneSmiskol/op_params)

Parameters are stored at `/data/op_params.json`

## opEdit Demo
<img src=".media/op_edit.gif?raw=true" width="1000">

Automatic updates
-----
When a new update is available on GitHub for Stock Additions, your EON/C2 will pull and reset your local branch to the remote. It then queues a reboot to occur when the following is true:
- your EON has been inactive or offroad for more than 5 minutes.

Therefore, if your device sees an update while you're driving it will reboot approximately 5 to 10 minutes after you finish your drive, it resets the timer if you start driving again before the time is up.

Offline crash logging
-----
If you experience a crash or exception while driving with this fork and you're not on internet for the error to be uploaded to Sentry, you should be able to check out the directory `/data/community/crashes` to see any and all logs of exceptions caught in `manager.py`. Simply `cat` the log file you wish to view. This does not catch all errors, for example scons compilation errors or Python syntax errors will not be caught, `tmux a` is usually best to view these (if openpilot didn't start).

Documentation
=====

Quick Installation
-----
To install Stock Additions, just run the following on your EON/C2 (make sure to press enter after each line):

```
cd /data/
mv openpilot openpilot.old  # or equivalent
git clone -b stock_additions --single-branch https://github.com/shanesmiskol/openpilot --depth 1
reboot
```

The `--depth 1` flag shallow clones the fork, it ends up being about 90 Mb so you can get the fork up and running quickly. Once you install Stock Additions, [automatic updating](#Automatic-updates) should always keep openpilot up to date with the latest from my fork!

**NEW❗** Or use the [emu CLI](https://github.com/emu-sh/.oh-my-comma) to easily switch to this fork's default branch: `emu fork switch ShaneSmiskol`. This should get you up and running even quicker.

Branches
-----
Most of the branches on this fork are development branches I use as various openpilot tests. The few that more permanent are the following:
  * [`stock_additions`](https://github.com/ShaneSmiskol/openpilot/tree/stock_additions): This is similar to stock openpilot's release branch. Will receive occasional and tested updates to Stock Additions.
  * [`stock_additions-devel`](https://github.com/ShaneSmiskol/openpilot/tree/stock_additions-devel): My development branch of Stock Additions I use to test new features or changes; similar to the master branch. Not recommendeded as a daily driver.
  * [`traffic_lights-074`](https://github.com/ShaneSmiskol/openpilot/tree/traffic_lights-074):  The development branch testing our traffic light detection model [found here](https://github.com/ShaneSmiskol/traffic-lights).

## Archive Stock Additions branches
* [Stock Additions 0.7](https://github.com/ShaneSmiskol/openpilot/tree/stock_additions-07)
* [Stock Additions 0.7.1](https://github.com/ShaneSmiskol/openpilot/tree/stock_additions-071)
* [Stock Additions 0.7.4](https://github.com/ShaneSmiskol/openpilot/tree/stock_additions-074)
* [Stock Additions 0.7.5](https://github.com/ShaneSmiskol/openpilot/tree/stock_additions-075)

Videos
-----
Here's a short video showing how close the traffic profile was in `0.7.4`. In `0.7.5`, the traffic profile is an average of 7.371 feet closer from 18 mph to 90 mph. Video thanks to [@rolo01](https://github.com/rolo01)!

[![](https://img.youtube.com/vi/sGsODeP_G_c/0.jpg)](https://www.youtube.com/watch?v=sGsODeP_G_c)


RELEASES.md
Stock Additions 0.7.7
========================
 * LaneSpeed process is much more efficient, CPU usage down from 6% (causing lag alerts) to ~1%
   * Use numpy_fast.interp instead of np.interp for slight boost in performance
 * Support white panda with experimental parameter. localizer performance may be reduced
 * Fixed silent alerts due to the new alert manager in 0.7.7
 * Make all the UI buttons show their state with unique colors!
 * Add derivative to lateral PI control for TSS1 Corolla and Prius!

Version 0.7.7 (2020-07-20)
========================
 * White panda is no longer supported, upgrade to comma two or black panda
 * Improved vehicle model estimation using high precision localizer
 * Improved thermal management on comma two
 * Improved autofocus for road-facing camera
 * Improved noise performance for driver-facing camera
 * Block lane change start using blindspot monitor on select Toyota, Hyundai, and Subaru
 * Fix GM ignition detection
 * Code cleanup and smaller release sizes
 * Hyundai Sonata 2020 promoted to officially supported car
 * Hyundai Ioniq Electric Limited 2019 and Ioniq SE 2020 support thanks to baldwalker!
 * Subaru Forester 2019 and Ascent 2019 support thanks to martinl!

Version 0.7.6.1 (2020-06-16)
========================
 * Hotfix: update kernel on some comma twos (orders #8570-#8680)

Version 0.7.6 (2020-06-05)
========================
 * White panda is deprecated, upgrade to comma two or black panda
 * 2017 Nissan X-Trail, 2018-19 Leaf and 2019 Rogue support thanks to avolmensky!
 * 2017 Mazda CX-5 support in dashcam mode thanks to Jafaral!
 * Huge CPU savings in modeld by using thneed!
 * Lots of code cleanup and refactors

Stock Additions 0.7.5
========================
 * Various opEdit improvements, colors, highlighting of last explored param, quicker messages
 * New Dynamic Follow modification to help slow and accelerate sooner
 * Live parameters update quicker
 * dfManager reliability improvements
 * New LaneSpeed and Dynamic Camera Offset features added

Version 0.7.5 (2020-05-13)
========================
 * Right-Hand Drive support for both driving and driver monitoring!
 * New driving model: improved at sharp turns and lead speed estimation
 * New driver monitoring model: overall improvement on comma two
 * Driver camera preview in settings to improve mounting position
 * Added support for many Hyundai, Kia, Genesis models thanks to xx979xx!
 * Improved lateral tuning for 2020 Toyota Rav 4 (hybrid)

Stock Additions 0.7.4 (version 0.2)
========================
 * Sidebar will not pop out when you tap the Dynamic Follow profile change button when driving
 * Add derivative to longcontrol. Improves responsiveness and helps overshoot

Version 0.7.4 (2020-03-20)
========================
 * New driving model: improved lane changes and lead car detection
 * Improved driver monitoring model: improve eye detection
 * Improved calibration stability
 * Improved lateral control on some 2019 and 2020 Toyota Prius
 * Improved lateral control on VW Golf: 20% more steering torque
 * Fixed bug where some 2017 and 2018 Toyota C-HR would use the wrong steering angle sensor
 * Support for Honda Insight thanks to theantihero!
 * Code cleanup in car abstraction layers and ui

Version 0.7.3 (2020-02-21)
========================
 * Support for 2020 Highlander thanks to che220!
 * Support for 2018 Lexus NX 300h thanks to kengggg!
 * Speed up ECU firmware query
 * Fix bug where manager would sometimes hang after shutting down the car

Version 0.7.2 (2020-02-07)
========================
 * ECU firmware version based fingerprinting for Honda & Toyota
 * New driving model: improved path prediction during turns and lane changes and better lead speed tracking
 * Improve driver monitoring under extreme lighting and add low accuracy alert
 * Support for 2019 Rav4 Hybrid thanks to illumiN8i!
 * Support for 2016, 2017 and 2020 Lexus RX thanks to illumiN8i!
 * Support for 2020 Chrysler Pacifica Hybrid thanks to adhintz!

Stock Additions 0.7.1
========================
 * Recover faster when lead leaves path, or when braking or acceleration is required immediately. Also should speed up the acceleration with auto lane change.
 * Add 3 different dynamic follow profiles: `roadtrip`, `relaxed`, and `traffic`. `relaxed` is the current dynamic follow profile. You can also swap profiles without rebooting live by using opEdit. SSH in and: `cd /data/openpilot;python op_edit.py` The parameter to change is `dynamic_follow`
 * Fix for steering unavailable when wheel goes over 100 degrees/sec.
 * Tuning for dynamic gas
 * Accelerate quicker and get closer when you're lane changing
 * Higher acceleration, and higher limits for turns
 * Automatic updates, EON will reboot by itself as long as it is inactive

Version 0.7.1 (2020-01-20)
========================
 * comma two support!
 * Lane Change Assist above 45 mph!
 * Replace zmq with custom messaging library, msgq!
 * Supercombo model: calibration and driving models are combined for better lead estimate
 * More robust updater thanks to jyoung8607! Requires NEOS update
 * Improve low speed ACC tuning

Stock Additions 0.7
========================
 * Dynamic lane speed is a new feature that reduces your cruising speed if many vehicles around you are significantly slower than you. This works with and without an openpilot-identified lead.
 * Dynamic gas tuning. Above 20 mph we take lead velocity and the following distance into account. Possibility of different tuning for different cars in the future. (DYNAMIC GAS NOW ONLY WORKS ON TOYOTA COROLLA AND RAV4 PEDAL)
 * Dynamic follow tuning, don't get as close when lead is accelerating.
 * Added static_steer_ratio parameter, if True openpilot will use the steer ratio in your interface file. Default is true, false uses the openpilot learned value which can vary through your drives.
 * Added ability to live tune parameters with `op_tune.py`. Currently only the camera offset (`camera_offset`) is supported.

selfdrive/eta.py
import datetime
import numpy as np
import time
import threading
import os


class ETAData:
  def __init__(self, t=0, progress=0):
    self.time = t
    self.progress = progress

class ETA(threading.Thread):
  def __init__(self, start_time, max_progress, frequency, sfp, spinner=None):  # only supports up to minutes
    self.start_time = start_time
    self.max_progress = max_progress
    self.frequency = frequency
    self.time = 0
    self.etr = 0  # in seconds, estimated time remained
    self._start = time.time()
    self.progress_subtract = 0
    self.times = [727.05, 722.89, 722.78, 722.72, 722.69, 716.83, 713.04, 706.94, 705.69, 704.11, 704.09, 700.19, 699.27, 698.14, 698.12, 689.12, 687.45, 685.81, 684.36, 682.18, 679.99, 679.14, 677.8, 675.98, 675.24, 673.88, 672.43, 670.69, 669.88, 668.05, 667.24, 665.81, 664.14, 663.32, 661.78, 660.33, 659.36, 658.12, 656.63, 655.63, 653.92, 652.04, 651.31, 649.84, 648.61, 647.47, 646.18, 644.82, 643.74, 642.48, 640.7, 639.51, 638.31, 636.98, 635.88, 634.72, 632.98, 632.25, 630.78, 629.21, 627.47, 626.57, 626.55, 626.54, 626.53, 626.52, 623.98, 613.96, 613.73, 613.53, 609.79, 586.08, 586.06, 586.06, 586.05, 586.04, 586.03, 585.94, 576.13, 572.26, 571.6, 570.47, 568.58, 568.35, 568.34, 568.33, 568.32, 568.31, 568.3, 568.3, 567.93, 566.71, 563.85, 563.03, 562.28, 559.7, 559.01, 559.0, 558.99, 558.98, 558.97, 558.14, 557.55, 557.54, 557.53, 557.52, 557.51, 557.5, 557.49, 557.49, 557.45, 555.27, 554.51, 554.47, 552.79, 551.8, 551.78, 547.81, 547.77, 546.45, 545.48, 545.47, 545.46, 545.45, 544.42, 543.46, 543.45, 543.44, 542.35, 541.26, 540.21, 539.04, 538.2, 532.34, 530.35, 529.73, 529.35, 528.5, 528.44, 528.43, 527.33, 526.22, 525.13, 523.84, 521.69, 506.09, 504.01, 503.22, 502.32, 502.31, 502.3, 502.29, 502.28, 502.27, 502.26, 502.25, 502.24, 502.24, 502.23, 502.22, 502.21, 502.2, 502.19, 502.18, 502.17, 502.16, 502.15, 502.14, 502.13, 502.12, 502.11, 502.1, 502.09, 502.09, 502.07, 502.06, 502.06, 502.05, 502.04, 502.03, 502.02, 502.01, 502.0, 501.99, 501.98, 501.97, 501.96, 501.95, 501.94, 501.93, 489.7, 276.17, 275.48, 243.8, 214.41, 200.68, 161.13, 147.59, 143.19, 131.22, 130.58, 130.57, 130.56, 128.82, 128.74, 128.3, 110.79, 110.78, 110.6, 110.59, 110.58, 110.57, 110.56, 110.55, 101.57, 93.07, 85.54, 84.73, 57.15, 45.54, 45.46, 31.75, 29.88, 15.1, 13.76, 13.27, 12.45, 8.48, 0.0]
    self.seconds = 60
    self.total_ips = 0
    self.spinner = spinner
    self.this_ips = 0
    self.last_ips = 0
    self.updated = False
    self.etr = 0
    self.run_thread = False
    self.can_reset = False
    self.eta_data = []
    self.scons_finished_progress = sfp
    threading.Thread.__init__(self)
    self.window_len = 20 * self.frequency  # in seconds * frequency
    # self.etrs = [12.15 * 60 for _ in range(int(self.window_len / 2))]
    self.etrs = []
    self.ipss = []
    # self.ipss = [1.95 for _ in range(int(self.window_len / 2))]

  # def init(self, t, max_progress):
  #     self.start_time = t
  #     self.max_progress = max_progress

  def get_eta_data(self, idx=-1):
    # By default this will return latest
    if len(self.eta_data) == 0:
      return ETAData()
    return self.eta_data[idx]

  def run(self):
    progress = 0
    while progress < self.max_progress:
      progress = self.get_eta_data().progress
      if os.path.exists('/data/stop'):
        os.remove('/data/stop')
        return
      if not self.run_thread:
        self.spinner.update("%d" % (progress / self.max_progress * self.scons_finished_progress), 'calculating...')
        time.sleep(1)
        continue

      eta_message = self.get_eta()
      self.spinner.update("%d" % (progress / self.max_progress * self.scons_finished_progress), eta_message)
      # spinner.update("%d" % (percentage * scons_finished_progress), eta_message)
      if self.updated:
        self.updated = False
      time.sleep(1 / self.frequency)

  def update(self, progress, t):
    self.can_reset = True
    self.last_ips = float(self.this_ips)
    self.eta_data.append(ETAData(progress=progress, t=t))
    self.updated = True
    removed = False
    while len(self.eta_data) > 3:  # we only need past 3
      del self.eta_data[0]
      removed = True

    if time.time() - self.get_eta_data().time < 1e-4:  # we don't care about these updates
      self.progress_subtract += 1
    # else:
    #   print('GOT UPDATE')
    print('data len: {}'.format(len(self.eta_data)))
    if not self.run_thread:
      self.run_thread = removed  # wait until we have enough data

  def moving_average(self, seq):
    w = np.hanning(self.window_len)

    s = np.r_[seq[self.window_len-1:0:-1], seq, seq[-1:-self.window_len:-1]]
    y = np.convolve(w / w.sum(), s, mode='valid')
    return y[self.window_len]

  def set_ips(self):
    # print(self.time)
    # print(self.last_time)
    cur_time = time.time()
    # # if self.has_update:
    self.this_ips = (self.get_eta_data().progress - self.get_eta_data(-2).progress) / (cur_time - self.get_eta_data(-2).time)
    print('last ips: {}'.format(self.last_ips))
    if self.this_ips < 10 < self.last_ips and self.can_reset:  # and self.updated:
      self.start_time = cur_time  # reset total ips when we stop getting cached files
      self.progress_subtract = int(self.get_eta_data(-1).progress)
      print('RESET HERE! p_s: {}, p: {}\n---------'.format(self.progress_subtract, int(self.get_eta_data(-1).progress)))
      self.total_ips = self.this_ips
    else:
      # print('progress: {}'.format(self.get_eta_data().progress))
      # print('time: {}\n---'.format(cur_time - self.start_time))
      self.total_ips = (self.get_eta_data().progress - self.progress_subtract) / (cur_time - self.start_time)

  def keep_len(self):
    while len(self.etrs) > self.window_len:
      del self.etrs[0]
    while len(self.ipss) > self.window_len:
      del self.ipss[0]

  def get_eta(self):
    self.keep_len()
    self.set_ips()
    percentage = round(self.get_eta_data().progress / self.max_progress * 100, 1)
    print('TOTAL IPS: {}\n---------'.format(round(self.total_ips, 2)))
    while len(self.etrs) > self.window_len:
      del self.etrs[0]

    # ips = self.total_ips * 0.8 + self.this_ips * 0.2
    ips = self.this_ips

    # x = np.linspace(0, 1, 10)
    # y = [(i + 1) ** -2 for i in x]
    #
    # ips = np.interp(ips / 10, x, y) * ips  # penalize large ips's
    self.ipss.append(ips)
    if len(self.ipss) - 1 < self.window_len:
      # print('calculated: {}'.format(y[-1]))
      return "calculating step 1..."

    ips = self.moving_average(self.ipss)

    self.etr = (self.max_progress - self.get_eta_data().progress) / ips
    self.etrs.append(self.etr)


    if len(self.etrs) - 1 < self.window_len:
      return "calculating step 2..."

    etr = self.format_etr(self.moving_average(self.etrs))


    # ips *= np.interp(self.get_eta_data().progress, [self.scons_finished_progress * 0.75, self.scons_finished_progress], [1.0, 0.5])

    # if self.this_ips < ips:
    #   ips = self.this_ips * 0.2 + ips * 0.8
    #   if self.last_ips < ips:
    #     ips = self.last_ips * 0.2 + ips * 0.8
    print('USING IPS: {} THIS IPS: {}\n---------'.format(round(ips, 2), round(self.this_ips, 2)))

    # self.etr -= ips / self.frequency




    # avg = self.total_ips * 0.7 + self.this_ips * 0.15 + self.last_ips * 0.15
    # elif ips < 10:
    #   self.etr -= ips / self.frequency
    #   # print('LAST UPDATE OVER 5 SECONDS!')
    #   # return 'compiling: {}% ETA: {}'.format(percentage, self.format_etr(self.etr))
    # print('NORMAL ETR: {}'.format(self.etr))

    # etr = self.format_etr(y[self.window_len])

    # # print('after etr: {}'.format(self.etr))
    # etr = self.format_etr(self.etr)

    return 'compiling: {}% ETA: {}'.format(percentage, etr)
  def format_etr(self, etr):
    hours, remainder = divmod(round(etr), self.seconds ** 2)
    minutes, seconds = divmod(remainder, self.seconds)

    time_list = [hours, minutes, seconds]
    time_str_list = ['hour', 'minute', 'second']

    etr_list = []
    for t, t_str in zip(time_list, time_str_list):
      plural = 's' if t != 1 else ''
      if t != 0:
        etr_list.append('{} {}{}'.format(int(t), t_str, plural))
    return ', '.join(etr_list)
